exec_test.go

```diff
@@ -494,3 +493,105 @@ func TestCommandCombinedOutput(t *testing.T) {
 		})
 	}
 }
+
+func TestCommandStdin(t *testing.T) {
+	d := newTestDistro(t, jammyRootFs)
+
+	const (
+		readFromPipe int = iota + 1
+		readFromBuffer
+	)
+
+	testCases := map[string]struct {
+		text string // We'll write this to Stdin. A python program will read it back to us.
+		// It is therefore both input and part of the expected output.
+
+		closeBeforeWait bool // Set to true to close the pipe before execution of the Cmd is over
+		readFrom        int  // Where Cmd should read Stdin from
+	}{
+		"standard":         {text: "Hello, wsl!", readFrom: readFromPipe},
+		"funny characters": {text: "Hello, \x00\twsl!", readFrom: readFromPipe},
+		"closing early":    {text: "Hello, wsl!", closeBeforeWait: true, readFrom: readFromPipe},
+		"using a buffer":   {text: "Hello, wsl!", readFrom: readFromBuffer},
+	}
+
+	// Simple program to test stdin
+	command := `python3 -c '
+from time import sleep
+v = input("Write your text here: ")
+sleep(1)					        # Ensures we get the prompts in separate reads
+print("Your text was", v)
+'`
+
+	for name, tc := range testCases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+			defer cancel()
+
+			os.Setenv("WSL_UTF8", "1")
+			cmd := d.Command(ctx, command)
+
+			out, err := cmd.StdoutPipe()
+			require.NoError(t, err, "Failed to pipe stdout")
+			cmd.Stderr = cmd.Stdout
+
+			var stdin io.Writer
+			switch tc.readFrom {
+			case readFromPipe:
+				stdin, err = cmd.StdinPipe()
+				require.NoError(t, err, "Failed to pipe stdin")
+			case readFromBuffer:
+				stdinbuff := bytes.NewBufferString(tc.text + "\n")
+				cmd.Stdin = stdinbuff
+				stdin = stdinbuff
+			default:
+				t.Fatalf("test implementation error: unrecognized testCase.readFrom: %d", tc.readFrom)
+			}
+
+			_, err = cmd.StdinPipe()
+			require.Error(t, err, "Unexpected success calling (*Cmd).StdinPipe twice")
+
+			err = cmd.Start()
+			require.NoError(t, err, "Unexpected error calling (*Cmd).Start")
+
+			defer cmd.Wait() // nolint: errcheck
+			// We know it's going to fail in normal execution. It's here to close
```

Comment:
Shouldn’t the comment be before the call to be clear on what it applies?

Reply:
Good call. I also rewrote it because the text is a bit unclear.